src/Nethermind/Nethermind.Xdc/EpochSwitchManager.cs

```diff
@@ -0,0 +1,454 @@
+        }
+
+        ulong closeEpochNum = 2ul;
+
+        if (closeEpochNum >= epochRound - tempTCEpoch)
+        {
+            var estBlockHeader = (XdcBlockHeader)_tree.FindHeader(estBlockNum);
+            if (estBlockHeader is null)
+            {
+                epochBlockInfo = null;
+                return false;
+            }
+            var epochSwitchInfos = GetEpochSwitchBetween(estBlockHeader, headHeader);
+            if (epochSwitchInfos is null)
+            {
+                return false;
+            }
+            foreach (var info in epochSwitchInfos)
+            {
+                ulong epochNum = (ulong)xdcSpec.SwitchEpoch + info.EpochSwitchBlockInfo.Round / (ulong)xdcSpec.EpochLength;
+                if (epochNum == tempTCEpoch)
+                {
+                    epochBlockInfo = info.EpochSwitchBlockInfo;
+                    return true;
+                }
+            }
+        }
+
+        return TryBinarySearchBlockByEpochNumber(tempTCEpoch, estBlockNum, epochSwitchInfo.EpochSwitchBlockInfo.BlockNumber, (ulong)xdcSpec.SwitchBlock, (ulong)xdcSpec.EpochLength, out epochBlockInfo);
+    }
+
+    private bool TryGetBlockInfoInCache(ulong estRound, ulong epoch, out BlockInfo epochBlockInfo)
+    {
+        var epochSwitchInCache = new List<BlockInfo>();
+
+        for (ulong r = estRound; r < estRound + (ulong)epoch; r++)
+        {
+            if (_round2EpochBlockInfo.TryGetValue(r, out BlockInfo blockInfo) && blockInfo is not null)
+            {
+                epochSwitchInCache.Add(blockInfo);
+            }
+        }
+
+        if (epochSwitchInCache.Count == 1)
+        {
+            epochBlockInfo = epochSwitchInCache[0];
+            return true;
+        }
+        else if (epochSwitchInCache.Count == 0)
+        {
+            epochBlockInfo = null;
+            return false;
+        }
+
+        foreach (var blockInfo in epochSwitchInCache)
+        {
+            var header = _tree.FindHeader(blockInfo.BlockNumber);
+            if (header is null)
+            {
+                continue;
+            }
+            if (header.Hash == blockInfo.Hash)
+            {
+                epochBlockInfo = blockInfo;
+                return true;
+            }
+        }
+
+        epochBlockInfo = null;
+        return false;
+    }
+
+    private bool TryBinarySearchBlockByEpochNumber(ulong tempTCEpoch, long start, long end, ulong switchBlock, ulong epoch, out BlockInfo epochBlockInfo)
+    {
+        while (start < end)
+        {
+            var header = (XdcBlockHeader)_tree.FindHeader((start + end) / 2);
+            if (header is null)
+            {
+                epochBlockInfo = null;
+                return false;
+            }
+
+            bool isEpochSwitch = IsEpochSwitchAtBlock(header, out ulong epochNum);
+
+            if (epochNum == tempTCEpoch)
+            {
+                if (!Utils.TryGetExtraFields(header, (long)switchBlock, out var extraFieldsV2, out _))
```

Comment:
Why use this one here if you only care about the round? And in similar situations above you'd do `if (header.ExtraConsensusData is null)` instead? Just saying in order to keep it consistent